### PR TITLE
Improvements to the driver

### PR DESCRIPTION
**Pull request by dvanaken**
- Added support for Oracle v12.1c.
- The Oracle collector now prepends the view name to each metric to avoid overwriting metrics from different views with the same names.
- Updated the knob/metric output samples for each of the Oracle versions we support since my fix changed all of them.
- The driver now works when the database system host is local, remote, or running in a Docker container.
- Changed the format of the driver config from JSON to a python settings file to make it easier to update / read.
- Removed old LHS code from driver (which is now supported by the website)

### CODE REVIEW

File: client/driver/fabfile.py [hunk 1]
```diff
@@ -89,64 +73,90 @@ def check_disk_usage():
...
-        cmd = 'sh oracleScripts/shutdownOracle.sh && sh oracleScripts/startupOracle.sh'
+    if dconf.DB_TYPE == 'postgres':
+        if dconf.HOST_CONN == 'docker':
+            # Restarting the docker container here is the cleanest way to do it
+            # becaues there's no init system running and the only process running
+            # in the container is postgres itself
+            local('docker restart {}'.format(dconf.CONTAINER_NAME))
+        else:
+            sudo('pg_ctl -D {} -w -t 600 restart -m fast'.format(
+                dconf.PG_DATADIR), user=dconf.DB_USER)
```
Comment (yangdsh) [inline client/driver/fabfile.py]: `DB_USER` should be linux_user, where the linux user is usually `postgres`.

File: client/driver/fabfile.py [hunk 2]
```diff
@@ -252,41 +266,43 @@ def save_dbms_result():
...
     return t
 
 
 @task
 def free_cache():
-    cmd = 'sync; sudo bash -c "echo 1 > /proc/sys/vm/drop_caches"'
-    local(cmd)
+    if dconf.HOST_CONN != 'docker':
+        with show('everything'), settings(warn_only=True):  # pylint: disable=not-context-manager
+            res = sudo("sync && echo 3 | tee /proc/sys/vm/drop_caches")
```
Comment (yangdsh) [inline client/driver/fabfile.py]: On my machine. This command causes permission deny.
I need to use `res = sudo("sh -c \"echo 3 > /proc/sys/vm/drop_caches\"")`

File: client/driver/fabfile.py [hunk 3]
```diff
@@ -180,60 +193,61 @@ def change_conf(next_conf=None):
...
 
-    tmpconf = 'tmp_' + os.path.basename(CONF['database_conf'])
-    with open(tmpconf, 'w') as f:
+    tmp_conf_out = os.path.join(dconf.TEMP_DIR, os.path.basename(dconf.DB_CONF) + '.out')
+    with open(tmp_conf_out, 'w') as f:
         f.write(''.join(lines))
 
-    local('sudo cp {0} {0}.ottertune.bak'.format(CONF['database_conf']))
-    local('sudo mv {} {}'.format(tmpconf, CONF['database_conf']))
+    run('cp {0} {0}.ottertune.bak'.format(dconf.DB_CONF))
```
Comment (yangdsh) [inline client/driver/fabfile.py]: On my local machine, run this command causes permission deny.
sudo() works fine.

File: client/driver/fabfile.py [hunk 4]
```diff
@@ -180,60 +193,61 @@ def change_conf(next_conf=None):
...
 def signal_controller():
-    pidfile = os.path.join(CONF['controller_home'], 'pid.txt')
+    pidfile = os.path.join(dconf.CONTROLLER_HOME, 'pid.txt')
     with open(pidfile, 'r') as f:
         pid = int(f.read())
-    cmd = 'sudo kill -2 {}'.format(pid)
-    with lcd(CONF['controller_home']):  # pylint: disable=not-context-manager
+    cmd = 'kill -2 {}'.format(pid)
+    with lcd(dconf.CONTROLLER_HOME):  # pylint: disable=not-context-manager
         local(cmd)
```
Comment (yangdsh) [inline client/driver/fabfile.py]: On my local machine, run this command causes permission deny.
sudo() works fine.

File: client/driver/utils.py [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+@task
+def sudo(cmd, user=None, **kwargs):
+    if dconf.HOST_CONN == 'remote':
+        res = _sudo(cmd, user=user, **kwargs)
+
+    elif dconf.HOST_CONN == 'local':
+        pre_cmd = 'sudo '
+        if user:
+            pre_cmd += '-u {} '.format(user)
+        res = local(pre_cmd + cmd, capture=True, **kwargs)
```
Comment (yangdsh) [inline client/driver/utils.py]: `capture=True` causes postgres restart never completes.
Removing `capture=True` works fine on my machine for the whole pipeline.

File: client/driver/fabfile.py (hunk 4 above)
Comment (dvanaken) [reply client/driver/fabfile.py]: I think we discussed that this was due to how Gradle was installed on the ottertune-demo machine. Skipping this for now.

File: client/driver/utils.py (hunk 5 above)
Comment (dvanaken) [reply client/driver/utils.py]: Fabric's `local` command does not return anything (the stdout, stderr, return_code, etc.) when `capture` is False so I do not want to set it that way in most cases. For now I've added `capture` as an arg to the `sudo` and `run` functions in utils.py. By default it's set to True but I've set it to False in restart_database. If this still doesn't work let me know.

**Comment by dvanaken** [comment]
@yangdsh Thanks a bunch for testing this out locally.

**Pull request merged**